### PR TITLE
Client Server Conversation

### PR DESCRIPTION
**Pull request by Stanislava88**
Here i have solved the task for conversation between client and server

### CODE REVIEW

File: tests/com/clouway/clientserver/ClientTest.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                out.close();
+                socket.close();
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        }
+    }
+
+    private Client client;
+    private ServerSocket server;
```
Comment (freekman) [inline tests/com/clouway/clientserver/ClientTest.java]: this looks really strange to use ```private ServerSocket server;``` and in the test to make ```  FakeDateServer server = new FakeDateServer();```

File: tests/com/clouway/clientserver/ClientTest.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    @Rule
+    public JUnitRuleMockery context = new JUnitRuleMockery();
+
+    private Display display = context.mock(Display.class);
+
+    class FakeDateServer extends Thread {
+        @Override
+        public void run() {
+            try {
+                server = new ServerSocket(2024);
```
Comment (freekman) [inline tests/com/clouway/clientserver/ClientTest.java]: .

File: tests/com/clouway/clientserver/ClientTest.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                ServerSocket server = new ServerSocket(2024);
+                Socket socket = server.accept();
+
+                OutputStreamWriter out = new OutputStreamWriter(socket.getOutputStream());
+                out.write("Some text");
+
+                out.close();
+                socket.close();
+                server.close();
+            } catch (IOException e) {
```
Comment (freekman) [inline tests/com/clouway/clientserver/ClientTest.java]: just asking if you use this aproach will you have to 
```js
    try( ServerSocket socket = new ServerSocket(5555)){
      
    }catch (IOException r){}
```
do this 
```js
 socket.close();
 server.close();
```

File: src/com/clouway/clientserver/Clock.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+import java.util.Date;
+
+/**
+ * The implementation of this interface will be used to get time
+ *
+ * @author Stanislava Kaukova(sisiivanovva@gmail.com)
+ */
+public interface Clock {
+    /**
+     * Will return Date
```
Comment (freekman) [inline src/com/clouway/clientserver/Clock.java]: ``` Will return Date current Date ```

File: src/com/clouway/clientserver/DateServer.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import java.io.OutputStreamWriter;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.Date;
+
+/**
+ * @author Stanislava Kaukova(sisiivanovva@gmail.com)
+ */
+public class DateServer extends Thread {
+    private int port;
```
Comment (freekman) [inline src/com/clouway/clientserver/DateServer.java]: final

File: src/com/clouway/clientserver/DateServer.java [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+    public DateServer(int port, Clock clock) {
+        this.port = port;
+        this.clock = clock;
+    }
+
+    @Override
+    public void run() {
+        try {
+            serverSocket = new ServerSocket(port);
```
Comment (freekman) [inline src/com/clouway/clientserver/DateServer.java]: leave it like it is but just to know you could use the server/client naming
```js
   server = new ServerSocket(port);
   client = serverSocket.accept();
```

File: src/com/clouway/clientserver/DateServer.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+            Socket socket = serverSocket.accept();
+
+            OutputStreamWriter out = new OutputStreamWriter(socket.getOutputStream());
+            Date date = clock.getTime();
```
Comment (freekman) [inline src/com/clouway/clientserver/DateServer.java]: clock get time to return Date does not soud good why not  ```Date date = clock.getDate();```

File: src/com/clouway/clientserver/DateServer.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+            out.write("Hello!" + date );
+
+            out.close();
+            socket.close();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+    }
+
+    public void close() throws IOException {
```
Comment (freekman) [inline src/com/clouway/clientserver/DateServer.java]: if the client forgets to close the Server what will happen

File: tests/com/clouway/clientserver/ServerTest.java [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+        context.checking(new Expectations() {{
+            oneOf(clock).getTime();
+            will(returnValue(date));
+        }});
+
+        String actual = client.connect();
+        String expected = "Hello!" + date;
+
+        System.out.println(expected);
```
Comment (freekman) [inline tests/com/clouway/clientserver/ServerTest.java]: remove

File: tests/com/clouway/clientserver/ClientTest.java (hunk 3 above)
Comment (Stanislava88) [inline tests/com/clouway/clientserver/ClientTest.java]: If i use this i don't need method close()

File: src/com/clouway/clientserver/DateServer.java (hunk 8 above)
Comment (Stanislava88) [inline src/com/clouway/clientserver/DateServer.java]: the resources that were allocated for this connection don't  be released

**Comment by freekman** [comment]
LGTM

**Pull request merged**